d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingState.java

```diff
@@ -0,0 +1,121 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+
+
+public class SubsettingState
+{
+  private final Object _lock = new Object();
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>> _weightedSubsetsCache;
+
+  @GuardedBy("_lock")
+  private long _version;
+
+  @GuardedBy("_lock")
+  private long _peerClusterVersion;
+
+  @GuardedBy("_lock")
+  private long _minClusterSubsetSize;
+
+  public SubsettingState(SubsettingStrategyFactory subsettingStrategyFactory)
+  {
+    _version = -1;
+    _peerClusterVersion = -1;
+    _minClusterSubsetSize = -1;
+    _subsettingStrategyFactory = subsettingStrategyFactory;
+    _weightedSubsetsCache = new HashMap<>();
+  }
+
+  public long getPeerClusterVersion(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return -1;
+    }
+
+    return subsettingStrategy.getPeerClusterVersion();
+  }
+
+  public Map<URI, TrackerClient> getClientsSubset(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId,
+      Map<URI, TrackerClient> potentialClients,
+      long version)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return potentialClients;
+    }
+
+    long peerClusterVersion = subsettingStrategy.getPeerClusterVersion();
+
+    synchronized (_lock)
+    {
+      if (version == _version &&
+          peerClusterVersion == _peerClusterVersion &&
+          minClusterSubsetSize == _minClusterSubsetSize &&
```

Comment:
If I am understanding this code correctly, could we put these three conditions behind a method called `isCacheValid()` or something like that?

Reply:
Thanks for reviewing the code! Appreciate your suggestions!

Agreed isCacheValid() would be cleaner